src/core/regimes.rkt

```diff
@@ -76,22 +77,33 @@
   (define prog-body (program-body prog))
   (for/list ([expr (remove-duplicates (subexprs-in-expr prog-body))]
              #:when (and (not (null? (free-variables expr)))
+                         (equal? (type-of expr (for/list ([var (program-variables prog)]) (cons var 'real))) 'real)
                          (critical-subexpression? prog-body expr)))
     expr))
 
-(define (combine-alts best-option)
+(define (combine-alts best-option precision)
   (match-define (option splitindices alts pts expr _) best-option)
   (match splitindices
    [(list (si cidx _)) (list-ref alts cidx)]
    [_
-    (define splitpoints (sindices->spoints pts expr alts splitindices))
+    (define splitpoints (sindices->spoints pts expr alts splitindices precision))
     (debug #:from 'regimes "Found splitpoints:" splitpoints ", with alts" alts)
 
     (define expr*
       (for/fold
           ([expr (program-body (alt-program (list-ref alts (sp-cidx (last splitpoints)))))])
           ([splitpoint (cdr (reverse splitpoints))])
-        `(if (<= ,(sp-bexpr splitpoint) ,(sp-point splitpoint))
+        (define prec-point (match precision
+                             [(or 'binary64 'binary32) (sp-point splitpoint)]
+                             ['posit8 `(real->posit8 ,(posit8->double (sp-point splitpoint)))]
+                             ['posit16 `(real->posit16 ,(posit16->double (sp-point splitpoint)))]
+                             ['posit32 `(real->posit32 ,(posit32->double (sp-point splitpoint)))]))
```

Comment:
Do I understand correctly, that we're still doing `sindices→spoints` in double precision, and only converting to posits here? That seems like the wrong way round, especially since the range of the two number systems is different.

Reply:
The `sindices->spoints` is being done in posits (through `midpoint-float` in `float.rkt`, which is a bad name, but I think all the renaming can happen with the transition to the interface.) However, the displayed value isn't quite right since it is being displayed as the nearest double number rather than the exact posit number. This something that will be fixed with the number system interface because we will require that users specify conversions to arbitrary-precision values of their numbers, so we can actually print the value of the splitpoint to whatever accuracy we want. I think any accuracy lost through not getting exactly the right splitpoint is pretty minimal here, but we gain a lot in actually being able to read the value of the splitpoint.